dockerd panic: slice bounds out of range in raft transport.splitSnapshotData (snapshot send)

## What happened

We hit a recurring dockerd panic while operating a Swarm cluster (panic inside swarmkit raft snapshot transport). When a manager needs to send a raft snapshot (observed while joining/promoting additional managers), the leader crashes with a slice-bounds panic in `transport.splitSnapshotData`.

The panic repeats on restart while the node is still participating in the raft cluster.

## Environment

- Docker Engine: 29.2.1 (linux)
- SwarmKit module version from stack trace: `github.com/moby/swarmkit/v2@v2.1.2-0.20251110192100-17b8d222e7dd`
- Linux (systemd)

## Trigger / suspected conditions

- Observed when adding/promoting additional managers (leader sending raft snapshot to a peer).
- In our case the swarm state had grown quite large (we observed ~1858 swarm `config` objects at one point, mostly generated by automation; only a small fraction were referenced by services). We removed the unused configs, but still saw the panic until rebuilding the swarm from scratch.

## Expected

Sending a snapshot to a peer should not crash the manager.

## Actual

Leader panics and exits; systemd restarts dockerd; the panic may repeat.

## Crash excerpt

```
panic: runtime error: slice bounds out of range [:9065848] with capacity 4871629
goroutine 161470 [running]:
github.com/moby/swarmkit/v2/manager/state/raft/transport.splitSnapshotData(...)
        github.com/moby/swarmkit/v2@v2.1.2-0.20251110192100-17b8d222e7dd/manager/state/raft/transport/peer.go:177 +0x265
github.com/moby/swarmkit/v2/manager/state/raft/transport.(*peer).sendProcessMessage(...)
        github.com/moby/swarmkit/v2@v2.1.2-0.20251110192100-17b8d222e7dd/manager/state/raft/transport/peer.go:246 +0x22f
github.com/moby/swarmkit/v2/manager/state/raft/transport.(*peer).run(...)
        github.com/moby/swarmkit/v2@v2.1.2-0.20251110192100-17b8d222e7dd/manager/state/raft/transport/peer.go:420 +0x1d4
created by github.com/moby/swarmkit/v2/manager/state/raft/transport.newPeer
        github.com/moby/swarmkit/v2@v2.1.2-0.20251110192100-17b8d222e7dd/manager/state/raft/transport/peer.go:63 +0x332
```

## Workaround

Rebuilding the swarm from scratch avoided the crash for us.

If you’d like, I can try to reproduce in a minimal environment and provide more details (snapshot size, exact steps).

--- a/manager/state/raft/transport/peer.go
+++ b/manager/state/raft/transport/peer.go
@@ -172,6 +172,10 @@ func splitSnapshotData(_ context.Context, m *raftpb.Message) []api.StreamRaftMes
 		}
 
 		raftMsg := *m
+		// Clone Snapshot so that re-slicing Snapshot.Data below
+		// does not mutate m.Snapshot.Data through the shared pointer.
+		snap := *m.Snapshot
+		raftMsg.Snapshot = &snap
 
 		// sub-slice for this snapshot chunk.
 		raftMsg.Snapshot.Data = m.Snapshot.Data[snapDataIndex : snapDataIndex+chunkSize]

--- a/manager/state/raft/transport/transport_test.go
+++ b/manager/state/raft/transport/transport_test.go
@@ -105,6 +105,51 @@ func testSend(ctx context.Context, c *mockCluster, from uint64, to []uint64, msg
 	}
 }
 
+// TestSplitSnapshotDataDoesNotMutateInput is a regression test for #3231.
+// Before the fix, splitSnapshotData did a shallow copy of the raft message
+// and re-sliced the shared Snapshot.Data on each iteration, shrinking the
+// original slice's capacity and eventually panicking with
+// "slice bounds out of range".
+func TestSplitSnapshotDataDoesNotMutateInput(t *testing.T) {
+	ctx := context.Background()
+
+	// Build a MsgSnap whose Snapshot.Data clearly exceeds GRPCMaxMsgSize so
+	// that the split loop runs multiple iterations (where the bug manifests).
+	const dataSize = 3 * GRPCMaxMsgSize
+	data := make([]byte, dataSize)
+	for i := range data {
+		data[i] = byte(i % (1 << 8))
+	}
+	m := raftpb.Message{
+		Type: raftpb.MsgSnap,
+		From: 1,
+		To:   2,
+		Snapshot: &raftpb.Snapshot{
+			Data: data,
+			Metadata: raftpb.SnapshotMetadata{
+				Index: uint64(len(data)),
+			},
+		},
+	}
+	origData := m.Snapshot.Data
+	origLen, origCap := len(origData), cap(origData)
+
+	msgs := splitSnapshotData(ctx, &m)
+	require.Greater(t, len(msgs), 1, "data larger than GRPCMaxMsgSize must split into multiple chunks")
+
+	// Chunks must reassemble to the original data.
+	var assembled []byte
+	for _, msg := range msgs {
+		assembled = append(assembled, msg.Message.Snapshot.Data...)
+	}
+	assert.Equal(t, data, assembled)
+
+	// The input message's Snapshot.Data must be untouched (regression guard).
+	assert.Equal(t, origLen, len(m.Snapshot.Data))
+	assert.Equal(t, origCap, cap(m.Snapshot.Data))
+	assert.Equal(t, data, m.Snapshot.Data)
+}
+
 func TestSend(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	c := newCluster()

